eclair-core/src/main/scala/fr/acinq/eclair/PluginParams.scala

```diff
@@ -55,4 +54,18 @@ trait CustomCommitmentsPlugin extends PluginParams {
    * returned by this method.
    */
   def getHtlcsRelayedOut(htlcsIn: Seq[IncomingHtlc], nodeParams: NodeParams, log: LoggingAdapter): Map[Origin, Set[(ByteVector32, Long)]]
+}
+
+/**
+ * Each intercepted message type must be handled by at most one plugin.
+ */
+object InterceptedMessageType extends Enumeration {
+  type InterceptedMessageType = Value
+  val InterceptOpenChannel: InterceptedMessageType.Value = Value(1, "InterceptOpenChannel")
+  val InterceptUpdateAddHtlc: InterceptedMessageType.Value = Value(2, "InterceptUpdateAddHtlc")
+}
+
+trait InterceptMessagePlugin extends PluginParams {
+  // set of message types that can be intercepted by this plugin
+  def canIntercept: Set[InterceptedMessageType.Value]
```

Comment:
This is premature factoring, it looks like you're trying to build a _generic_ way of intercepting arbitrary messages. That's not our goal since every interception mechanism will require very custom work and different mechanisms inside eclair to make them possible. The only known cases we'll want to provide interception for are channel opening and htlc forwarding.

Keep it simple, just define separate traits for each interception case:

```
trait InterceptChannelOpenPlugin extends PluginParams
(definitely not in this PR)
trait InterceptHtlcsPlugin extends PluginParams
```